### PR TITLE
Implemented a fix for the setTimeout unsubscribe thing

### PR DESCRIPTION
**Pull request by Poincare**
Should hopefully be a fix for the #1409 and some other related issues.

Took me a while to work out what kinds of tests broke and it's pretty likely that this change breaks behavior that we don't have tests for so we should release it with fingers crossed :)

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [x] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

### CODE REVIEW

**Comment by helfer** [comment]
@Poincare I'm a bit hesitant merging this, but we could release it under a `next` tag and ask as many people as we can to try it out. Before we do that though, I'd like to be sure that it leads to a significant performance improvement. Can you run some in-browser benchmarks or profiles and then report back?

**Comment by Poincare** [comment]
@helfer 

Safari:
dope

Chrome:
even more dope

:) :)

**Comment by helfer** [comment]
Ok, remind me again, what was the baseline?

File: src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -103,18 +103,29 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
...
     return new Promise((resolve, reject) => {
-      const subscription = this.subscribe({
+      let subscription: (Subscription | null) = null;
+      const observer: Observer<ApolloQueryResult<T>> = {
         next(result) {
           resolve(result);
+
+          // Stop the query within the QueryManager if we can before
+          // this function returns.
+          const selectedObservers = that.observers.filter((obs: Observer<ApolloQueryResult<T>>) => obs !== observer);
```
Comment (helfer) [inline src/core/ObservableQuery.ts]: Is this filter here really necessary, or could we just check if the length is 1?

Comment (Poincare) [inline src/core/ObservableQuery.ts]: This filter is necessary due to some weird race condition that another part of the code mentions. I'm essentially matching the code in `tearDownQuery`.

**Comment by Poincare** [comment]
Mean time for the baseline on Chrome, the problematic browser, was 192.4 ms ([see here](https://github.com/apollographql/apollo-client/issues/1409)).

File: src/core/ObservableQuery.ts [hunk 2]
```diff
@@ -103,18 +103,29 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
...
+    const that = this;
     return new Promise((resolve, reject) => {
-      const subscription = this.subscribe({
+      let subscription: (Subscription | null) = null;
+      const observer: Observer<ApolloQueryResult<T>> = {
         next(result) {
           resolve(result);
+
+          // Stop the query within the QueryManager if we can before
+          // this function returns.
```
Comment (stubailo) [inline src/core/ObservableQuery.ts]: I think we should add more comments about why we need this.

**Pull request merged**